department-of-veterans-affairs/notification-api
# #747 VEText Migration - Update Twilio provider logic to use sms sender. Branch - feature/747-2

# Description

POST requests to /v2/notifications/sms are used to send text messages to veterans.  In addition to the recipient's phone number or recipient_id, which can be used to look up the phone number, sending a request requires the sender's phone number.

These changes implement the ability to create text notifications using the value of *message_service_sid* in place of the sender's phone number when using Twilio as the SMS backend.  *messaging_service_sid*, if available, will be read from the *sms_sender_specifics* JSON field of a ServiceSmsSender instance associated with the authenticated service (user) and given template id, which is the other required field when making a request to this sms endpoint.

I found the existing code very difficult to follow.  Many of the edits in this PR are comments I made about types and purposes to facilitate not having to refer to different files frequently.  I also made formatting changes to appease the code scan test.

## Rough flow path for the endpoint
1. app/v2/notifications/post_notifications.py::post_notification
2. The above calls either *process_sms_or_email_notification* or *process_notification_with_recipient_identifier*.  The outcome is the creation of a new Notification instance from the JSON attributes contained in the POST request.
3. The notification is sent to *send_notification_to_queue* where it's added to the Celery Queue. 
4. The Celery worker processes each task by passing it to the "provider" (i.e. Twilio, etc.).  That is implemented in app/delivery/send_to_providers.py::send_sms_to_provider.

The `sms_sender_id` is also passed along the above flow so the specific one can be chosen when it's time to send the message.

Fixes #747

## Type of change
Please delete options that are not relevant.


## How Has This Been Tested?

- [ ] Test sending an sms with the VEText service using twilio as the provider on the sms template. You can use the following information to test it in Dev while this branch is deployed:
### Testing Twilio with Postman
1. use this POST request: `/v2/notifications/send SMS/send sms with phone_number`
2. use your phone number for the `phone_number` field to make sure the messages are going through
3. use the following information to make the request

#### VEText Service Info
Use the VEText service ID

Create or obtain an API key for the service

Use the following sms-template-id:
bec7cdd1-2934-424d-a27d-f5f9e7c24bb5

sms-sender-id:
8451a581-5316-4402-8bfb-70633ee2fb24 - no `sender_specifics`
c458d3c2-8442-4d6f-be4b-8811cc11a847 - with `sender_specifics`
^ these are to test with and without including `sender_specifics` in the `sms_sender`

### Setup steps to pass sms through VEText on Dev
- Added sms template to VEText service
- Added Twilio `provider_id` to the VEText sms template
- Added `sms-sender` with `sms_sender_specifics`
  - used this value for twilio ```"sms_sender_specifics": {
            "messaging_service_sid": "MG[redacted]"
        }```
- Added api key for VEText on Dev 
- Set Twilio provider `active=true` on Dev 
- ~~Changed parameter store (ssm) value:~~ `/dev/notification-api/twilio/from-number`
  - The value is no longer used. I reset it to the old value. 
    - The number in the `sms-sender: sms_sender` is used for the `from_number` value when sending using the Twilio provider
  - new value: +19703005680 - works
  - old value: +16173263357 - errors

## Checklist
